Foreign key between reference tables

pykello opened the pull request:
DESCRIPTION: Support Foreign Key constraints between reference tables.

The changes are:
1. Modify `ErrorIfUnsupportedForeignConstraint()` to not error out for foreign keys for reference tables.
2. Modify `SerializeNonCommutativeWrites()` to lock referenced (and transitively referenced) tables to avoid concurrency issues like https://github.com/citusdata/citus/pull/2864#issuecomment-520020075
3. Update `foreign_key_restriction_enforcement.sql` with tests for dist -> ref -> ref foreign keys. Behaviour should be similar to dist -> ref case.
4. Add `isolation_ref2ref_foreign_keys_on_mx.spec` to make sure that we are getting the correct locks in the first worker to avoid replica divergence in MX.

marcocitus commented:
Do we need any changes to the locking? e.g. if an update cascades from reference table A to reference table B while reference table B is concurrently being updated, could the updates on table B be applied in a different order on different replicas?

pykello commented:
> Do we need any changes to the locking? e.g. if an update cascades from reference table A to reference table B while reference table B is concurrently being updated, could the updates on table B be applied in a different order on different replicas?

Just to document a case that doesn't work without the locks:

Preparation:

```sql
create table ref(a int primary key, b int);
select create_reference_table('ref');
create table t(a int, b int);
select create_reference_table('t');
alter table t add constraint tfk foreign key (b) references ref(a) on update cascade;
insert into ref values (1, 1), (3, 3);
insert into t values (1, 1), (3, 3);
```

1.sql:
```sql
update ref set a = case when a % 2 == 1 then a + 1 else a - 1 end;
```

2.sql:
```
update t set a = (a + 1) % 1000 where b = 4;
```

Concurrently run 1.sql and 2.sql using pgbench. 2.sql will produce warnings like `WARNING:  modified 1 tuples, but expected to modify 0`, and after pgbench is done, we see that the replicas of `t` have diverged.

I'll add the necessary locks to avoid this.

pykello commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -938,13 +956,6 @@ ExecuteSingleSelectTask(CitusScanState *scanState, Task *task)
 			placementAccessList = list_make1(placementAccess);
 		}
 
-		if (placementAccessList == NIL)
```
Not possible, because of preceding checks.

marcocitus replied:
The error below suggests a condition under which it is possible (FindShardPlacementOnGroup in BuildPlacementAccessList returns NULL because a shard move completed in between planning and execution). We should remove `Assert(list_length(placementAccessList) == list_length(relationShardList));` instead.

pykello commented on src/backend/distributed/executor/multi_router_executor.c:
```diff
@@ -1316,9 +1327,6 @@ GetModifyConnections(Task *task, bool markCritical)
 	}
 
-	/* then finish in parallel */
-	FinishConnectionListEstablishment(multiConnectionList);
```
We are not doing async connection establishment here.

pykello commented on src/test/regress/expected/foreign_key_restriction_enforcement.out:
```diff
@@ -249,6 +420,35 @@ DETAIL:  Reference relation "reference_table" is modified, which might lead to d
+	SELECT count(*) FROM on_update_fkey_table WHERE value_1 = 101 AND id = 99;
+ count 
+-------
+     0
```
Maybe we need to create the tables in such a way that changes to `transitive_reference_table` cascades to `on_update_fkey_table` ...

marcocitus commented on src/backend/distributed/utils/resource_lock.c:
```diff
@@ -336,21 +338,58 @@ LockReferencedReferenceShardDistributionMetadata(uint64 shardId, LOCKMODE lock)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		LockShardResource(shardInterval->shardId, lockMode);
+	}
```
would it make sense to use `LockShardListResources` here or did you avoid it to skip the extra sort?

pykello replied:
> would it make sense to use `LockShardListResources` here or did you avoid it to skip the extra sort?

Yeah, just wanted to skip the extra sort.

marcocitus commented on src/backend/distributed/commands/foreign_constraint.c:
```diff
@@ -150,65 +150,65 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
+			 * If the table is just being distributed, IsDistributedTable() returns
+			 * false, so we need a special check for self-referencing tables.
+			 */
+			referencedIsReferenceTable = (distributionMethod == DISTRIBUTE_BY_NONE);
```
could we fill in the distributionMethod at the start if the table already exists to avoid the extra call to PartitionMethod above?

pykello replied:
> could we fill in the distributionMethod at the start if the table already exists to avoid the extra call to PartitionMethod above?

`PartitionMethod` was called once per referenced relation. Since this function was very confusing, I did a total overhaul of it after your comment. Trying to make things a bit more obvious. The logic is divided into three functions:

1. ErrorIfUnsupportedForeignConstraintExists -> Does the high level looping and information preparation.
2. ErrorIfUnsupportedForeignConstraint -> Our main logic
3. ForeignConstraintFindDistKeys -> A low-level detail which took too much space, so I thought it can be refactored out.

The logic is same as before, but in 3 separate functions.

marcocitus commented on src/backend/distributed/utils/resource_lock.c:
```diff
@@ -534,12 +573,21 @@ SerializeNonCommutativeWrites(List *shardIntervalList, LOCKMODE lockMode)
-		LockShardListResourcesOnFirstWorker(lockMode, shardIntervalList);
+		if (ClusterHasKnownMetadataWorkers() && !IsFirstWorkerNode())
+		{
+			LockShardListResourcesOnFirstWorker(lockMode, shardIntervalList);
```
are we doing this twice now? (again in LockReferencedReferenceShardResources)

pykello replied:
> are we doing this twice now? (again in LockReferencedReferenceShardResources)

Here it is for the relation itself, in `LockReferencedReferenceShardResources` it is for referenced relations. I am changing the variable names there to avoid confusion.

marcocitus commented on src/test/regress/expected/foreign_key_restriction_enforcement.out:
```diff
@@ -249,6 +420,35 @@ DETAIL:  Reference relation "reference_table" is modified, which might lead to d
 
+ROLLBACK;
+BEGIN;
+	UPDATE transitive_reference_table SET id = 101 WHERE id = 99;
```
was this meant to update id or value_1?

marcocitus commented on src/test/regress/specs/isolation_ref2ref_foreign_keys_on_mx.spec:
```diff
@@ -0,0 +1,153 @@
+
+	-- start_metadata_sync_to_node can not be run inside a transaction block
+	-- following is a workaround to overcome that
+	-- port numbers are hard coded at the moment
```
could use punctuation :)

marcocitus commented:
Do we need non-MX isolation tests as well?

pykello commented:
> Do we need non-MX isolation tests as well?

Yes. working on it. I'll add them soon.

pykello commented on src/backend/distributed/commands/foreign_constraint.c:
```diff
@@ -149,237 +150,291 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
+	 * We support foreign keys between reference tables. No more checks
+	 * are necessary.
+	 */
+	if (referencingIsReferenceTable && referencedIsReferenceTable)
```
Except this line + removal of the related error, everything else in this file is just moving things around and renaming variables so code can be understood a bit easier.

pykello commented:
> > Do we need non-MX isolation tests as well?
> 
> Yes. working on it. I'll add them soon.

Added non-MX isolation tests and fixed a cache invalidation related bug. We need to guarantee `MetadataCache.distColocationRelationId` is initialized to be able to get foreign key graph invalidation messages. Previously this could be 0 when invalidation messages arrived, hence the issue.

marcocitus commented:
Minor issues with our current implementation of locking in case of foreign keys that do are beyond the scope of this PR:
- A multi-shard update/delete on a distributed table with a foreign key to a reference table could result in a deadlock if there is a concurrent update on the reference table. The solution would be for the update on the reference table to take multi-shard locks.
- Shard moves can always trigger a deadlock because transactions can take metadata locks on co-located shards and reference tables in an arbitrary order, while the shard move takes metadata locks on all shards in the foreign key graph

pykello commented on src/backend/distributed/commands/foreign_constraint.c:
```diff
@@ -149,237 +150,291 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
-									  "table")));
+			ereport(ERROR, (errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+							errmsg("cannot create foreign key constraint"),
+							errdetail("Referenced table must be a distributed table.")));
```
done

pykello commented on src/backend/distributed/commands/foreign_constraint.c:
```diff
@@ -149,237 +150,291 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
+	 * placements always be in the same state (b) executors are aware of reference
+	 * tables and handle concurrency related issues accordingly.
+	 */
+	if (!referencingSingleReplicated)
```
done

pykello commented on src/backend/distributed/commands/foreign_constraint.c:
```diff
@@ -149,237 +150,291 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
+									char referencedDistMethod,
+									Var *referencedDistColumn,
+									uint32 referencedColocationId,
+									HeapTuple pgConstraintTuple)
```
inlined this function. I think renaming the variables and reordering the code a bit is enough for making the code more understandable.

pykello commented on src/test/regress/specs/isolation_ref2ref_foreign_keys.spec:
```diff
@@ -0,0 +1,114 @@
+# Case 5. Modifying ref_table_2 should also lock ref_table_1 in RowExclusive mode.
+permutation "s2-begin" "s2-insert-table-2" "s1-view-locks" "s2-rollback" "s1-view-locks"
+# Case 6. Modifying ref_table_2 should also lock ref_table_1 in RowExclusive mode.
+permutation "s2-begin" "s2-insert-table-3" "s1-view-locks" "s2-rollback" "s1-view-locks"
```
Added some more tests.

pykello commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -2637,6 +2637,14 @@ InitializeCaches(void)
+			 * See the comments of InvalidateForeignKeyGraph for more context.
+			 */
+			MetadataCache.distColocationRelationId =
+				get_relname_relid("pg_dist_colocation", PG_CATALOG_NAMESPACE);
```
I need to think about this a bit more. It seems that this `InitializeCaches()` can be called even before creation of extension in which case `DistPartitionRelationId()` will error out. What I added will set `distColocationRelationId` to `InvalidOid`, I am not sure if this is ok. I just need to think more.

marcocitus replied:
Could we do this in CitusHasBeenLoaded?

pykello commented on src/backend/distributed/utils/metadata_cache.c:
```diff
@@ -3089,7 +3097,11 @@ ResetDistTableCacheEntry(DistTableCacheEntry *cacheEntry)
+	/*
+	 * We shouldn't call DistColocationRelationId() here since cache
+	 * invalidation can happen before postgres' relation cache being
+	 * initialized, in which case get_relname_relid() fails.
```
this comment is not accurate. I'll update it after spending time on understanding things.

marcocitus commented on src/backend/distributed/commands/foreign_constraint.c:
```diff
@@ -149,132 +155,110 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
-									  "table")));
+			ereport(ERROR, (errcode(ERRCODE_INVALID_TABLE_DEFINITION),
+							errmsg("cannot create foreign key constraint"),
+							errdetail("Referenced table must be a distributed table"
```
maybe "a distributed table or a reference table"

pykello replied:
> maybe "a distributed table or a reference table"

` a reference table` part is in the next line
